diff --git a/crypto/poly1305/asm/poly1305-x86_64.pl b/crypto/poly1305/asm/poly1305-x86_64.pl
index fa9bfb7a7b..51c1745da6 100755
--- a/crypto/poly1305/asm/poly1305-x86_64.pl
+++ b/crypto/poly1305/asm/poly1305-x86_64.pl
@@ -2814,7 +2814,6 @@ poly1305_blocks_vpmadd52:
 
 	shl	\$40,$padbit
 	mov	64($ctx),%r8			# peek on power of the key
-
 	# if powers of the key are not calculated yet, process up to 3
 	# blocks with this single-block subroutine, otherwise ensure that
 	# length is divisible by 2 blocks and pass the rest down to next
@@ -3383,7 +3382,6 @@ poly1305_blocks_vpmadd52_8x:
 
 	shl	\$40,$padbit
 	mov	64($ctx),%r8			# peek on power of the key
-
 	vmovdqa64	.Lx_mask44(%rip),$mask44
 	vmovdqa64	.Lx_mask42(%rip),$mask42
 
@@ -3395,6 +3393,21 @@ poly1305_blocks_vpmadd52_8x:
 	vmovq	16($ctx),%x#$H2
 
 .Lblocks_vpmadd52_8x:
+___
+$code.=<<___	if ($win64);
+	sub		\$0xf0,%rsp
+	vmovdqa		%xmm6,0x50(%rsp)
+	vmovdqa		%xmm7,0x60(%rsp)
+	vmovdqa		%xmm8,0x70(%rsp)
+	vmovdqa		%xmm9,0x80(%rsp)
+	vmovdqa		%xmm10,0x90(%rsp)
+	vmovdqa		%xmm11,0xa0(%rsp)
+	vmovdqa		%xmm12,0xb0(%rsp)
+	vmovdqa		%xmm13,0xc0(%rsp)
+	vmovdqa		%xmm14,0xd0(%rsp)
+	vmovdqa		%xmm15,0xe0(%rsp)
+___
+$code.=<<___;
 	################################################################
 	# fist we calculate more key powers
 
@@ -3730,7 +3743,21 @@ $code.=<<___;
 	vmovq		%x#$H1,8($ctx)
 	vmovq		%x#$H2,16($ctx)
 	vzeroall
-
+___
+$code.=<<___	if ($win64);
+	vmovdqa		0x50(%rsp),%xmm6
+	vmovdqa		0x60(%rsp),%xmm7
+	vmovdqa		0x70(%rsp),%xmm8
+	vmovdqa		0x80(%rsp),%xmm9
+	vmovdqa		0x90(%rsp),%xmm10
+	vmovdqa		0xa0(%rsp),%xmm11
+	vmovdqa		0xb0(%rsp),%xmm12
+	vmovdqa		0xc0(%rsp),%xmm13
+	vmovdqa		0xd0(%rsp),%xmm14
+	vmovdqa		0xe0(%rsp),%xmm15
+	add		\$0xf0,%rsp
+___
+$code.=<<___;
 .Lno_data_vpmadd52_8x:
 	ret
 .cfi_endproc
